pkg/apis/job/v1/job.proto

```diff
@@ -0,0 +1,203 @@
+  //
+  // The output always starts from the beginning of process execution, and will
+  // be written to the stream in real-time until the job completes, or until the
+  // stream is cancelled by the client.
+  //
+  // Jobs that were stopped by the user with the Stop() method will have output
+  // up to the time they were stopped.
+  //
+  // If the job is already completed, the full output of the job will be
+  // written to the stream, after which the stream will be closed.
+  rpc Output(JobId) returns (stream ProcessOutput) {
+    option (rbac.v1.scope).enabled = true;
+  }
+}
+
+// JobSpec describes a command to be run, along with optional resource limits
+// that should be applied to the command's process.
+message JobSpec {
+  CommandSpec    command = 1;
+  ResourceLimits limits  = 2;
+}
+
+message JobId {
+  string id = 1;
+}
+
+message JobIdList {
+  repeated JobId items = 1;
+}
+
+// State describes the logical state of a job.
+//
+//   ┌─────────────────────────────────────────┐
+//   │                                         │
+//   │              ┌────────┐                 │
+//   │      ┌──────►│ Failed │                 │
+//   │      │       └────────┘                 │
+//   │ ┌────┴────┐              ┌────────────┐ │
+//   │ │ Pending │              │ Terminated │ │
+//   │ └────┬────┘              └────────────┘ │
+//   │      │       ┌─────────┐        ▲       │
+//   │      └──────►│ Running ├────────┘       │
+//   │              └─────────┘                │
+//   │                                         │
+//   └─────────────────────────────────────────┘
+enum State {
+  Unknown = 0;
+  // The job is waiting to be started, and is not yet running.
+  Pending = 1;
+  // The job failed to start.
+  Failed = 2;
+  // The job started successfully and is currently running.
+  Running = 3;
+  // The job is no longer running.
+  Terminated = 4;
+}
+
+message JobStatus {
+  // The job's logical state.
+  State state = 1;
+  // The job's original spec.
+  JobSpec spec = 2;
+  // A human-readable message describing the job's state, or an error message
+  // if the job failed.
+  string message = 3;
+
+  // The PID of the job's process. Only present if the job is in the Running
+  // or Terminated state.
+  int32 pid = 4;
+  // The time at which the job was started. Only present if the job is
+  // in the Running or Terminated state.
+  google.protobuf.Timestamp startTime = 5;
+  // Termination details. Only present if the job is in the Terminated state.
+  TerminationStatus terminated = 6;
+}
+
+// Contains details about the cause of the process's termination.
+message TerminationStatus {
+  // If the process exited normally, this will contain its exit code. Otherwise,
+  // if it was terminated by a signal, it will not be present.
+  optional int32 exitCode = 1;
+  // If the process was terminated by a signal, this will contain the signal
+  // number. Otherwise, it will not be present.
+  optional int32 signal = 2;
```

Comment:
Though I may disagree, I respect that you have a strong opinion about this and I'd be interested to pick your brain a bit more here:

My main driver behind choosing optional for `exitCode` is the way zero-valued fields are handled in `protoreflect` and by extension, codecs like `protojson` and `prototext`. For example, `jobctl status` can be instructed to output the status message using `protojson`, and the result is correct:
```
❯ bin/jobctl status d46b91fa29674aa8ae5a6d28b99eb624 -o json
{
  "state": "TERMINATED",
  "spec": {
    "command": {
      "command": "make",
      "args": [
        "-j",
        "64"
      ]
    },
    "limits": {
      "cpu": "1000"
    }
  },
  "message": "exit status 0",
  "pid": 666500,
  "startTime": "2023-12-18T16:10:59.653959508Z",
  "terminated": {
    "exitCode": 0,
    "time": "2023-12-18T17:35:51.958686719Z"
  }
}
```
Because ExitCode has presence, the formatter can distinguish between present-but-zero and not-present. Otherwise, it has no way to know. I could use `-1` to indicate field absence but that doesn't fix the formatting situation. 

We could write our own printer instead of using protojson, but protojson does some extra magic which is all locked behind internal packages (dont even get me started on that... :laughing:) so it is sometimes necessary if we actually do need to json-encode messages. If we need to interop with Javascript for example, the presence semantics of optional scalars in Go matches more closely with Javascript, where everything is nullable. There are options to print zero-valued fields, but that has its own set of problems. 

You are correct that according to Rules As Written, optional scalars in proto3 are considered an anti-pattern. However, I think this recommendation - and implicit presence in general - ended up being a mistake in the long run (to be fair, predicting the future is hard), and probably why Editions is reverting back to explicit presence by default.

Anyway, sorry for the wall of text :smile: It's of course not really a big deal in this case - I would be perfectly fine with your suggested change - but I am genuinely curious of your thoughts on this, in general.

Reply:
Why is it so important that we add a zero exit code to the protojson output? A nonzero exit code generally means that the process has failed, so it only being present as a field if there's been an error isn't that bad, I think.

I wasn't aware that protobuf editions are moving back to explicit presence, that's actually a nice change for once!

Reply:
To make it more easily machine-readable; consider:
`jobctl status <id> -o json | jq '.terminated.exitCode'`
will be "0" if the code is printed, otherwise "null".

The default formatter can be a more terse textual representation, but providing a json option is a pretty nice QoL feature for CLIs.

(I'm definitely bikeshedding here, but it's in the spirit of this challenge I think)

Reply:
You could argue that in a real product you're either going to have a specification for the json output that's decoupled from the details of the proto (because you don't want to break user CLI automation when you change internal protocol details), or you're going to say "the output is protojson from that protobuf message" and expect consumers of the output to deal with the consequences of that.

What happens if the job status says that the job has exited on its own but the exit code is not present? If you say "assume it's 0" then your presence information becomes just a protojson representation hack, if you say it's invalid then you're adding (more) representable invalid states to your data format. 🤔

Reply:
(btw feel free to ignore this whole thread for the purpose of going forward with the challenge)